## Packages

In [344]:
import pandas as pd             # Traitement des tableaux de données 

# Découvertes des données 

Dans un premier temps, nous chargeons le jeu de données et prenons le temps de regarder des informations sommaires concernant les métadonnées et les données de nos différents jeux :

### Caractéristiques 

In [345]:
BAAC_2012_carac = pd.read_csv("Data/TRAIN/BAAC-Annee-2012/caracteristiques_2012_.csv", encoding="latin1", sep=";")
BAAC_2012_carac = BAAC_2012_carac.drop(BAAC_2012_carac.columns[0], axis=1)

BAAC_2012_carac.head()

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col  com  \
0  201200000001  12     3    16  1930    5    2    1  1.0    6   11   
1  201200000002  12     9     1  2145    5    2    1  1.0    3   11   
2  201200000003  12    11    20  1815    5    2    1  1.0    5  670   
3  201200000004  12     1     7  1915    5    2    1  1.0    3   51   
4  201200000006  12     1    22  1800    5    2    1  1.0    6   51   

                     adr gps        lat      long  dep  
0  33 Rue Roger Salengro   M  5052928.0  293643.0  590  
1           route de DON   M  5053380.0  292738.0  590  
2     rue Anatole France   M  5054757.0  292033.0  590  
3         Rue d'ESTAIRES   M  5053315.0  280229.0  590  
4     34 rue de la Marne   M  5053665.0  280409.0  590

In [346]:
print(f"Le fichier des CARACTERISTIQUES a pour dimensions : {BAAC_2012_carac.shape[0]}x{BAAC_2012_carac.shape[1]}.")
print(f"Le nombres d'accidents unique est de : {BAAC_2012_carac['Num_Acc'].nunique()}.")

Le fichier des CARACTERISTIQUES a pour dimensions : 56025x16.
Le nombres d'accidents unique est de : 56025.


### Lieux

In [347]:
BAAC_2012_place = pd.read_csv("Data/TRAIN/BAAC-Annee-2012/lieux_2012_.csv", encoding="latin1", sep=";")
BAAC_2012_place = BAAC_2012_place.drop(BAAC_2012_place.columns[0], axis=1)

BAAC_2012_place.head()

Num_Acc  catr   voie  v1   v2  circ  nbv    pr    pr1  vosp  prof  \
0  201200000001     3   39.0 NaN  NaN   2.0  0.0   6.0  875.0   0.0   1.0   
1  201200000002     3   41.0 NaN  NaN   2.0  2.0  13.0  250.0   0.0   1.0   
2  201200000003     3   41.0 NaN    D   2.0  2.0  11.0  500.0   0.0   1.0   
3  201200000004     3  947.0 NaN  NaN   2.0  2.0   0.0  220.0   0.0   1.0   
4  201200000006     4    0.0 NaN  NaN   2.0  0.0   0.0    0.0   0.0   0.0   

   plan  lartpc  larrout  surf  infra  situ  env1  
0   1.0     0.0     72.0   1.0    0.0   1.0   0.0  
1   1.0     0.0     65.0   1.0    0.0   1.0  99.0  
2   3.0     0.0     77.0   1.0    0.0   1.0   0.0  
3   1.0     0.0     70.0   1.0    0.0   1.0   3.0  
4   1.0     0.0     67.0   1.0    6.0   1.0   3.0

In [348]:
print(f"Le fichier des LIEUX a pour dimensions : {BAAC_2012_place.shape[0]}x{BAAC_2012_place.shape[1]}.")
print(f"Le nombre d'accidents uniques est de : {BAAC_2012_place['Num_Acc'].nunique()}.")

Le fichier des LIEUX a pour dimensions : 56025x18.
Le nombre d'accidents uniques est de : 56025.


### Usagers

In [349]:
BAAC_2012_users = pd.read_csv("Data/TRAIN/BAAC-Annee-2012/usagers_2012_.csv", encoding="latin1", sep=";")
BAAC_2012_users = BAAC_2012_users.drop(BAAC_2012_users.columns[0], axis=1)

BAAC_2012_users.head()

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  201200000001    1.0     1     1     2     5.0  11.0   0.0   0.0    0.0   
1  201200000001    NaN     3     4     1     5.0   NaN   3.0   3.0    2.0   
2  201200000002    1.0     1     1     1     5.0  13.0   0.0   0.0    0.0   
3  201200000002    2.0     2     1     1     0.0  11.0   0.0   0.0    0.0   
4  201200000002    1.0     1     1     1     5.0  21.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1939.0     A01  
1   2008.0     A01  
2   1994.0     A01  
3   1992.0     A01  
4   1976.0     B02

In [350]:
print(f"Le fichier des USAGERS a pour dimensions : {BAAC_2012_users.shape[0]}x{BAAC_2012_users.shape[1]}.")
print(f"Le nombre d'accidents uniques est de : {BAAC_2012_users['Num_Acc'].nunique()}.")

Le fichier des USAGERS a pour dimensions : 124598x12.
Le nombre d'accidents uniques est de : 56025.


Contrairement aux précédents "*.csv*", on constate une dimension bien plus importante en terme de lignes. Cela peut facilement s'expliquer, en effet, il n'est pas anormal d'observer pour un seul accident se produisant dans un seul lieu, avec un seul ensemble de caractéristique de posséder plusieurs victimes. Ainsi, plusieurs victimes étant liées à un même accident viennent dupliquer les lignes liées à un unique accident.

### Véhicules

In [351]:
BAAC_2012_vehic = pd.read_csv("Data/TRAIN/BAAC-Annee-2012/vehicules_2012_.csv", encoding="latin1", sep=";")
BAAC_2012_vehic = BAAC_2012_vehic.drop(BAAC_2012_vehic.columns[0], axis=1)

BAAC_2012_vehic.head()

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  201200000001     0     7       0  0.0   1.0   7.0  16.0     A01
1  201200000002     0     7       0  0.0   2.0   3.0  16.0     A01
2  201200000002     0    33       0  0.0   2.0   8.0   1.0     B02
3  201200000003     0     7       0  0.0   2.0   1.0   1.0     C03
4  201200000003     0     7       0  0.0   2.0   8.0   1.0     B02

In [352]:
print(f"Le fichier des VEHICULES a pour dimensions : {BAAC_2012_vehic.shape[0]}x{BAAC_2012_vehic.shape[1]}.")
print(f"Le nombre d'accidents uniques est de : {BAAC_2012_vehic['Num_Acc'].nunique()}.")

Le fichier des VEHICULES a pour dimensions : 95115x9.
Le nombre d'accidents uniques est de : 56025.


De nouveaux, on constate une valeur différente en terme de ligne. Certes inférieure à celle des usagers mais supérieure aux deux autres. Cela peut s'expliquer par le fait que si un ou plusieurs véhicules sont engagés, il se peut qu'un ou plusieurs usagers soient concernés pour le même véhicules (on pensera notamment aux passagers). Cependant, il aurait été anormal de voir plus de véhicules que d'usagers (véhicules fantômes ?). Cela nous rassure donc sur la qualité des données.

### Premières Conclusions :

On observe un nombre d'accidents au nombre de **"*56025*"**. Cependant, il peut arriver d'avoir plusieurs véhicules impliqués (donnant donc **"*95115*"** lignes), ou encore plusieurs usagers accidentés (amenant **"*124598*"** cas d'usagers impactés).

Afin de pouvoir construire un jeu de données complet et lire de potentielles corrélations, il est nécessaire de songer à comment fusionner ces différents jeux.

L'objectif est de prédire la gravité d'un accident relatif aux usagers en disposant notamment d'informations sur le lieu, les caractéristiques, les véhicules impliqués. Une première tentative serait de construire le jeu de données autour des usagers. Pour cela, nous pouvons utiliser plusieurs variables pour lier les différents jeux de données. On pense notamment à : **"*Num_Acc*"**, ou encore **"*num_veh*"**. En effet, on pourra aisément relier les **"caractéristiques"** et le **"lieu"** de l'accident aux **"usagers"** à l'aide du **"*Num_Acc*"**. Pour ce qui est de lier les **"*véhicules*"** aux **"*usagers*"**, il faudra lier correctement l'accidenté avec le bon véhicule. Ainsi, la mise en relation conjointe des usagers avec ces deux caractéristiques est nécessaire. 

In [353]:
merged_data = pd.merge(BAAC_2012_users, BAAC_2012_vehic, on=['Num_Acc', 'num_veh'], how='left')

merged_data.head()

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  201200000001    1.0     1     1     2     5.0  11.0   0.0   0.0    0.0   
1  201200000001    NaN     3     4     1     5.0   NaN   3.0   3.0    2.0   
2  201200000002    1.0     1     1     1     5.0  13.0   0.0   0.0    0.0   
3  201200000002    2.0     2     1     1     0.0  11.0   0.0   0.0    0.0   
4  201200000002    1.0     1     1     1     5.0  21.0   0.0   0.0    0.0   

   an_nais num_veh  senc  catv  occutc  obs  obsm  choc  manv  
0   1939.0     A01     0     7       0  0.0   1.0   7.0  16.0  
1   2008.0     A01     0     7       0  0.0   1.0   7.0  16.0  
2   1994.0     A01     0     7       0  0.0   2.0   3.0  16.0  
3   1992.0     A01     0     7       0  0.0   2.0   3.0  16.0  
4   1976.0     B02     0    33       0  0.0   2.0   8.0   1.0

In [354]:
print(f"Le nouveau fichier des USAGERS a pour dimensions : {merged_data.shape[0]}x{merged_data.shape[1]}.")
print(f"Le nombre d'accidents uniques est de : {merged_data['Num_Acc'].nunique()}.")

Le nouveau fichier des USAGERS a pour dimensions : 124598x19.
Le nombre d'accidents uniques est de : 56025.


On observe que le merging a fonctionné. En effet, pour chaque usager, il a désormais **"*7*"** nouvelles variables (**"*9*"** variables au total sur **"*vehicules_2012_.csv*"** mais **"*usagers_2012_.csv*"** possèdent déjà **"*Num_Acc*"** et **"*num_veh*"** donc seulement **"*7*"** variables à ajouter). Cela nous fait désormais un total de **"*19*"** variables pour **"*124 598*"** lignes (usagers).

Désormais, nous souhaitons faire de même avec les **"*lieux*"** et les **"*caractéristiques*"**. Cependant, il sera plus facile de les merge étant donné qu'il suffit d'associer à l'individu les attributs des deux précédents jeux de données conformément au **"*Num_Acc*"**. 

In [355]:
whole_data = pd.merge(merged_data, BAAC_2012_carac, on=['Num_Acc'], how="left")
whole_data = pd.merge(whole_data, BAAC_2012_place, on=['Num_Acc'], how="left")

whole_data.head()

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  201200000001    1.0     1     1     2     5.0  11.0   0.0   0.0    0.0   
1  201200000001    NaN     3     4     1     5.0   NaN   3.0   3.0    2.0   
2  201200000002    1.0     1     1     1     5.0  13.0   0.0   0.0    0.0   
3  201200000002    2.0     2     1     1     0.0  11.0   0.0   0.0    0.0   
4  201200000002    1.0     1     1     1     5.0  21.0   0.0   0.0    0.0   

   ...    pr1 vosp  prof  plan  lartpc  larrout  surf  infra  situ  env1  
0  ...  875.0  0.0   1.0   1.0     0.0     72.0   1.0    0.0   1.0   0.0  
1  ...  875.0  0.0   1.0   1.0     0.0     72.0   1.0    0.0   1.0   0.0  
2  ...  250.0  0.0   1.0   1.0     0.0     65.0   1.0    0.0   1.0  99.0  
3  ...  250.0  0.0   1.0   1.0     0.0     65.0   1.0    0.0   1.0  99.0  
4  ...  250.0  0.0   1.0   1.0     0.0     65.0   1.0    0.0   1.0  99.0  

[5 rows x 51 columns]

In [356]:
print(f"Le fichier total des USAGERS a pour dimensions : {whole_data.shape[0]}x{whole_data.shape[1]}.")
print(f"Le nombre d'accidents uniques est de : {whole_data['Num_Acc'].nunique()}.")

Le fichier total des USAGERS a pour dimensions : 124598x51.
Le nombre d'accidents uniques est de : 56025.


A l'instar du merging précédent, on observe que l'on conserve le même nombre d'observations mais qu'on augmente le nombre de variables prédictives. Par acquit de conscience, on vient s'assurer par le calcul que la dimension associée aux variables prédictives est la bonne en comptant le nombre d'éléments uniques dans une liste contenant toutes les variables prédictives au sein des 4 jeux de données : 

In [357]:
columns_list = []

columns_list.extend(BAAC_2012_carac.columns)
columns_list.extend(BAAC_2012_place.columns)
columns_list.extend(BAAC_2012_users.columns)
columns_list.extend(BAAC_2012_vehic.columns)

# Print the number of unique elements
print(f"Le nombre de variables prédictives uniques depuis le début est de : {len(set(columns_list))}.")

Le nombre de variables prédictives uniques depuis le début est de : 51.


On a donc bien une cohérence sur la fusion des différents jeux de données et nous pouvons désormais nous pencher sur la structure des données.

Nous décidons de construire ce fichier afin de le traiter dans un prochain notebook :

In [1]:
whole_data.to_csv('BAAC_2012_full.csv', index=False)

NameError: name 'whole_data' is not defined

# Pré-Traitement des données

In [358]:
whole_data.head()

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  201200000001    1.0     1     1     2     5.0  11.0   0.0   0.0    0.0   
1  201200000001    NaN     3     4     1     5.0   NaN   3.0   3.0    2.0   
2  201200000002    1.0     1     1     1     5.0  13.0   0.0   0.0    0.0   
3  201200000002    2.0     2     1     1     0.0  11.0   0.0   0.0    0.0   
4  201200000002    1.0     1     1     1     5.0  21.0   0.0   0.0    0.0   

   ...    pr1 vosp  prof  plan  lartpc  larrout  surf  infra  situ  env1  
0  ...  875.0  0.0   1.0   1.0     0.0     72.0   1.0    0.0   1.0   0.0  
1  ...  875.0  0.0   1.0   1.0     0.0     72.0   1.0    0.0   1.0   0.0  
2  ...  250.0  0.0   1.0   1.0     0.0     65.0   1.0    0.0   1.0  99.0  
3  ...  250.0  0.0   1.0   1.0     0.0     65.0   1.0    0.0   1.0  99.0  
4  ...  250.0  0.0   1.0   1.0     0.0     65.0   1.0    0.0   1.0  99.0  

[5 rows x 51 columns]

### Num_Acc

Identifiant de l'accident. Valeur vitale pour joindre les données. Cependant, au-delà de la fusion des données, ce numéro n'apporte rien en matière de prédiction. En effet, il est générique et ne transmet pas d'information complémentaire. On se demande alors s'il ne vaudrait pas mieux le supprimer.

In [359]:
print(f"Le nombre d'observations avec une absence d'identifiant d'accidant est de : {whole_data['Num_Acc'].isna().sum()}.")
print(f"Le nombre d'observations unique d'identifiant d'accidant est de : {whole_data['Num_Acc'].nunique()}.")

whole_data.drop('Num_Acc', axis=1, inplace=True)

Le nombre d'observations avec une absence d'identifiant d'accidant est de : 0.
Le nombre d'observations unique d'identifiant d'accidant est de : 56025.


### place

Permet de situer la place occupée dans le véhicule par l'usager au moment de l'accident. Les informations concernants les positions sont trouvables au sein du fichier **"*description-des-bases-de-donnees-onisr-annees-2005-a-2020.pdf*"** notamment. Cette donnée est majeure car il existe certainement une relation entre la position des victimes et la gravité de l'accident. En effet, en fonction de la position, les dispositifs de sécurité ne sont pas les mêmes. On soupçonne donc une importante majeure.

In [341]:
possible_values = whole_data['place'].unique()
na_count = whole_data['place'].isna().sum()
obs_count = whole_data.shape[0]
rep_na = na_count / obs_count * 100

print(f"Les valeurs possibles sont : {possible_values}.")
print(f"Le nombre de valeurs manquantes est de : {na_count} pour un total de {obs_count} observations soit {rep_na}%.")

Les valeurs possibles sont : [ 1. nan  2.  3.  4.  8.  5.  6.  7.  9.].
Le nombre de valeurs manquantes est de : 11174 pour un total de 124598 observations soit 8.968041220565338%.


Etant donné qu'il est difficile de prédire une position manquante car cela dépend notamment du véhicule, il est donc décidé de supprimer les lignes manquantes. Cependant, il est nécessaire d'attendre la prochaine variable afin d'ajouter une condition. En effet, selon la documentation, **'*10*'** doit correspondre au piéton mais il n'apparaît pas. Ainsi, seules les lignes contenant une valeur **"*Nan*"** **et** qui ne matcheront pas la position de piéton seront supprimées.

### catu

Correspond à la catégorie d'usager. Autrement dit, s'il s'agit d'un conducteur, d'un passager ou bien d'un piéton. Il est important de disposer de ces informations en cas de corrélations entre le statut de la victime et son état. On peut soumettre l'hypothèse que ce statut influe sur les situations dans lesquelles les usagers sont impliqués notamment (difficile d'aller à 130 km/h en tant que piéton) :

In [360]:
# Supprime seulement si 'place' est Nan ET que ce n'est pas un piéton
whole_data = whole_data[~((whole_data['catu'] != 3) & whole_data['place'].isna())]

possible_values = whole_data['catu'].unique()
na_count = whole_data['catu'].isna().sum()
rep_na = na_count / obs_count * 100

print(f"Les valeurs possibles sont : {possible_values}.")
print(f"Le nombre de valeurs manquantes est de : {na_count} pour un total de {obs_count} observations soit {rep_na}%.")

Les valeurs possibles sont : [1 3 2].
Le nombre de valeurs manquantes est de : 0 pour un total de 113424 observations soit 0.0%.


### grav

Etant la colonne à prédire, il est évident que nous souhaitons conserver cette dernière. Nous nous penchons sur la qualité de cette dernière :

In [361]:
possible_values = whole_data['grav'].unique()
na_count = whole_data['grav'].isna().sum()
rep_na = na_count / obs_count * 100

print(f"Les valeurs possibles sont : {possible_values}.")
print(f"Le nombre de valeurs manquantes est de : {na_count} pour un total de {obs_count} observations soit {rep_na}%.")

Les valeurs possibles sont : [1 4 3 2].
Le nombre de valeurs manquantes est de : 0 pour un total de 113424 observations soit 0.0%.


Tout semble en bon état, les valeurs possibles respectent la documentation. Cependant, dans l'exercice de prédiction, il est demandé de convertir cette colonne en binaire (**'*0*'** pour les cas sans dangers [1, 4] et **'*1*'** pour les cas graves [2, 3]) : 

In [362]:
whole_data['grav'] = whole_data['grav'].replace({1: 0, 4: 0, 2: 1, 3: 1})

possible_values = whole_data['grav'].unique()

print(f"Les valeurs possibles sont : {possible_values}.")

Les valeurs possibles sont : [0 1].


### sexe

Certaines études ont pu démontrer des disparités dans les comportements de conduite entre les différents sexes. Nous conservons cette donnée dans l'éventualité où une corrélation existerait entre le sexe, le statut de conducteur et la gravité :

In [364]:
possible_values = whole_data['sexe'].unique()
na_count = whole_data['sexe'].isna().sum()
rep_na = na_count / obs_count * 100

print(f"Les valeurs possibles sont : {possible_values}.")
print(f"Le nombre de valeurs manquantes est de : {na_count} pour un total de {obs_count} observations soit {rep_na}%.")

Les valeurs possibles sont : [2 1].
Le nombre de valeurs manquantes est de : 0 pour un total de 113424 observations soit 0.0%.


### trajet